app/src/full/java/io/homeassistant/companion/android/sensors/CarInfoSensorManager.kt

```diff
@@ -0,0 +1,112 @@
+            stateClass = SensorManager.STATE_CLASS_MEASUREMENT,
+        )
+        private val batteryLevel = SensorManager.BasicSensor(
+            "car_info_battery_level",
+            "sensor",
+            R.string.basic_sensor_name_car_info_battery_level,
+            R.string.sensor_description_car_info_battery_level,
+            "mdi:car-battery",
+            unitOfMeasurement = "%",
+            stateClass = SensorManager.STATE_CLASS_MEASUREMENT,
+        )
+    }
+    private lateinit var latestContext: Context
+
+    override fun docsLink(): String {
+        return "https://companion.home-assistant.io/docs/core/sensors"
+    }
+
+    override val name: Int
+        get() = R.string.sensor_name_car_info
+
+    override suspend fun getAvailableSensors(context: Context): List<SensorManager.BasicSensor> {
+        return listOf(
+            fuelLevel, batteryLevel
+        )
+    }
+
+    override fun requiredPermissions(sensorId: String): Array<String> {
+        return arrayOf("com.google.android.gms.permission.CAR_FUEL")
+    }
+
+    @RequiresApi(Build.VERSION_CODES.P)
+    override fun requestSensorUpdate(
+        context: Context
+    ) {
+        latestContext = context
+        updateEnergy()
+    }
+
+    @RequiresApi(Build.VERSION_CODES.P)
+    private fun updateEnergy() {
+        if (!isEnabled(latestContext, fuelLevel) && !isEnabled(latestContext, batteryLevel))
+            return
+
+        val now = System.currentTimeMillis()
+        if (listenerLastRegistered + SensorManager.SENSOR_LISTENER_TIMEOUT < now && isListenerRegistered) {
+            Log.d(TAG, "Re-registering listener as it appears to be stuck")
+            HaCarAppService.carInfo?.removeEnergyLevelListener(this)
+            isListenerRegistered = false
+        }
+
+        if(HaCarAppService.carInfo != null) {
+            HaCarAppService.carInfo?.addEnergyLevelListener(latestContext.mainExecutor, this)
+            Log.d(TAG, "CarInfo sensor listener registered")
+            isListenerRegistered = true
+            listenerLastRegistered = now.toInt()
+        } else {
+            Log.d(TAG, "CarInfo not available")
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    override fun onCarDataAvailable(data: EnergyLevel) {
+        if (data.fuelPercent.status == CarValue.STATUS_SUCCESS) {
+            onSensorUpdated(
+                latestContext,
+                fuelLevel,
+                data.fuelPercent.value!!,
+                fuelLevel.statelessIcon,
+                mapOf()
+            )
+        }
+        if (data.batteryPercent.status == CarValue.STATUS_SUCCESS) {
+            onSensorUpdated(
+                latestContext,
+                batteryLevel,
+                data.batteryPercent.value!!,
+                batteryLevel.statelessIcon,
+                mapOf()
+            )
+        }
+        HaCarAppService.carInfo?.removeEnergyLevelListener(this)
+        Log.d(TAG, "CarInfo sensor listener unregistered")
```

Comment:
Please use some logger instead of Log.d. This allows you to have more control over the logging behavior and can be easier to configure than using Log directly.

Reply:
The project is not setup to use something other than `Log.*` at the moment

Reply:
ok, it is sad